File: PHP/One/Transforms/InferTypesPlugins/ResolveMethodCalls.php
<?php

namespace One\Transforms\InferTypesPlugins\ResolveMethodCalls;

use One\Transforms\InferTypesPlugins\Helpers\InferTypesPlugin\InferTypesPlugin;
use One\Ast\Expressions\Expression;
use One\Ast\Expressions\UnresolvedMethodCallExpression;
use One\Ast\Expressions\InstanceMethodCallExpression;
use One\Ast\Expressions\StaticMethodCallExpression;
use One\Ast\Expressions\IMethodCallExpression;
use One\Ast\AstTypes\ClassType;
use One\Ast\AstTypes\InterfaceType;
use One\Ast\AstTypes\AnyType;
use One\Ast\AstTypes\TypeHelper;
use One\Transforms\InferTypesPlugins\Helpers\GenericsResolver\GenericsResolver;
use One\Ast\References\ClassReference;
use One\Ast\References\StaticThisReference;
use One\Ast\Types\Class_;
use One\Ast\Types\IInterface;
use One\Ast\Types\Method;

class ResolveMethodCalls extends InferTypesPlugin {
    function __construct() {
        parent::__construct("ResolveMethodCalls");
        
    }
    
    protected function findMethod($cls, $methodName, $isStatic, $args) {
        $allBases = $cls instanceof Class_ ? array_values(array_filter($cls->getAllBaseInterfaces(), function ($x) { return $x instanceof Class_; })) : $cls->getAllBaseInterfaces();
        
        $methods = array();
        foreach ($allBases as $base)
            foreach ($base->methods as $m) {
                $minLen = count(array_values(array_filter($m->parameters, function ($p) { return $p->initializer === null; })));
                $maxLen = count($m->parameters);
                $match = $m->name === $methodName && $m->isStatic === $isStatic && $minLen <= count($args) && count($args) <= $maxLen;
                if ($match)
                    $methods[] = $m;
            }
        
        if (count($methods) === 0)
            throw new \OneLang\Error("Method '" . $methodName . "' was not found on type '" . $cls->name . "' with " . count($args) . " arguments");
        else if (count($methods) > 1) {
            // TODO: actually we should implement proper method shadowing here...
            $thisMethods = array_values(array_filter($methods, function ($x) use ($cls) { return $x->parentInterface === $cls; }));
            if (count($thisMethods) === 1)
                return $thisMethods[0];
            throw new \OneLang\Error("Multiple methods found with name '" . $methodName . "' and " . count($args) . " arguments on type '" . $cls->name . "'");
        }
        return $methods[0];
    }
    
    protected function resolveReturnType($expr, $genericsResolver) {
        $genericsResolver->collectFromMethodCall($expr);
        
        for ($i = 0; $i < count($expr->args); $i++) {
            // actually doesn't have to resolve, but must check if generic type confirm the previous argument with the same generic type
            $paramType = $genericsResolver->resolveType($expr->method->parameters[$i]->type, false);
            if ($paramType !== null)
                $expr->args[$i]->setExpectedType($paramType);
            $expr->args[$i] = $this->main->runPluginsOn($expr->args[$i]);
            $genericsResolver->collectResolutionsFromActualType($paramType, $expr->args[$i]->actualType);
        }
        
        if ($expr->method->returns === null) {
            $this->errorMan->throw("Method (" . $expr->method->parentInterface->name . "::" . $expr->method->name . ") return type was not specified or infered before the call.");
            return;
        }
        
        $expr->setActualType($genericsResolver->resolveType($expr->method->returns, true), true, $expr instanceof InstanceMethodCallExpression && TypeHelper::isGeneric($expr->object->getType()));
    }
    
    protected function transformMethodCall($expr) {
        if ($expr->object instanceof ClassReference || $expr->object instanceof StaticThisReference) {
            $cls = $expr->object instanceof ClassReference ? $expr->object->decl : ($expr->object instanceof StaticThisReference ? $expr->object->cls : null);
            $method = $this->findMethod($cls, $expr->methodName, true, $expr->args);
            $result = new StaticMethodCallExpression($method, $expr->typeArgs, $expr->args, $expr->object instanceof StaticThisReference);
            $this->resolveReturnType($result, new GenericsResolver());
            return $result;
        }
        else {
            $resolvedObject = $expr->object->actualType !== null ? $expr->object : $this->main->runPluginsOn($expr->object);
            $objectType = $resolvedObject->getType();
            $intfType = $objectType instanceof ClassType ? $objectType->decl : ($objectType instanceof InterfaceType ? $objectType->decl : null);
            
            if ($intfType !== null) {
                $method = $this->findMethod($intfType, $expr->methodName, false, $expr->args);
                $result = new InstanceMethodCallExpression($resolvedObject, $method, $expr->typeArgs, $expr->args);
                $this->resolveReturnType($result, GenericsResolver::fromObject($resolvedObject));
                return $result;
            }
            else if ($objectType instanceof AnyType) {
                $expr->setActualType(AnyType::$instance);
                return $expr;
            }
            else { }
            return $resolvedObject;
        }
    }
    
    function canTransform($expr) {
        return $expr instanceof UnresolvedMethodCallExpression && !($expr->actualType instanceof AnyType);
    }
    
    function transform($expr) {
        return $this->transformMethodCall($expr);
    }
}

File: CSharp/One/Transforms/InferTypesPlugins/ResolveMethodCalls.cs
using One.Ast;
using One.Transforms.InferTypesPlugins.Helpers;
using System.Collections.Generic;

namespace One.Transforms.InferTypesPlugins
{
    public class ResolveMethodCalls : InferTypesPlugin {
        public ResolveMethodCalls(): base("ResolveMethodCalls")
        {
            
        }
        
        protected Method findMethod(IInterface cls, string methodName, bool isStatic, Expression[] args)
        {
            var allBases = cls is Class class_ ? class_.getAllBaseInterfaces().filter(x => x is Class) : cls.getAllBaseInterfaces();
            
            var methods = new List<Method>();
            foreach (var base_ in allBases)
                foreach (var m in base_.methods) {
                    var minLen = m.parameters.filter(p => p.initializer == null).length();
                    var maxLen = m.parameters.length();
                    var match = m.name == methodName && m.isStatic == isStatic && minLen <= args.length() && args.length() <= maxLen;
                    if (match)
                        methods.push(m);
                }
            
            if (methods.length() == 0)
                throw new Error($"Method '{methodName}' was not found on type '{cls.name}' with {args.length()} arguments");
            else if (methods.length() > 1) {
                // TODO: actually we should implement proper method shadowing here...
                var thisMethods = methods.filter(x => x.parentInterface == cls);
                if (thisMethods.length() == 1)
                    return thisMethods.get(0);
                throw new Error($"Multiple methods found with name '{methodName}' and {args.length()} arguments on type '{cls.name}'");
            }
            return methods.get(0);
        }
        
        protected void resolveReturnType(IMethodCallExpression expr, GenericsResolver genericsResolver)
        {
            genericsResolver.collectFromMethodCall(expr);
            
            for (int i = 0; i < expr.args.length(); i++) {
                // actually doesn't have to resolve, but must check if generic type confirm the previous argument with the same generic type
                var paramType = genericsResolver.resolveType(expr.method.parameters.get(i).type, false);
                if (paramType != null)
                    expr.args.get(i).setExpectedType(paramType);
                expr.args.set(i, this.main.runPluginsOn(expr.args.get(i)));
                genericsResolver.collectResolutionsFromActualType(paramType, expr.args.get(i).actualType);
            }
            
            if (expr.method.returns == null) {
                this.errorMan.throw_($"Method ({expr.method.parentInterface.name}::{expr.method.name}) return type was not specified or infered before the call.");
                return;
            }
            
            expr.setActualType(genericsResolver.resolveType(expr.method.returns, true), true, expr is InstanceMethodCallExpression instMethCallExpr && TypeHelper.isGeneric(instMethCallExpr.object_.getType()));
        }
        
        protected Expression transformMethodCall(UnresolvedMethodCallExpression expr)
        {
            if (expr.object_ is ClassReference classRef || expr.object_ is StaticThisReference) {
                var cls = expr.object_ is ClassReference classRef2 ? classRef2.decl : expr.object_ is StaticThisReference statThisRef ? statThisRef.cls : null;
                var method = this.findMethod(cls, expr.methodName, true, expr.args);
                var result = new StaticMethodCallExpression(method, expr.typeArgs, expr.args, expr.object_ is StaticThisReference);
                this.resolveReturnType(result, new GenericsResolver());
                return result;
            }
            else {
                var resolvedObject = expr.object_.actualType != null ? expr.object_ : this.main.runPluginsOn(expr.object_);
                var objectType = resolvedObject.getType();
                var intfType = objectType is ClassType classType ? ((IInterface)classType.decl) : objectType is InterfaceType intType ? intType.decl : null;
                
                if (intfType != null) {
                    var lambdaField = intfType.fields.find(x => x.name == expr.methodName && x.type is LambdaType lambdType && lambdType.parameters.length() == expr.args.length());
                    if (lambdaField != null) {
                        var lambdaCall = new LambdaCallExpression(new InstanceFieldReference(expr.object_, lambdaField), expr.args);
                        lambdaCall.setActualType((((LambdaType)lambdaField.type)).returnType);
                        return lambdaCall;
                    }
                    
                    var method = this.findMethod(intfType, expr.methodName, false, expr.args);
                    var result = new InstanceMethodCallExpression(resolvedObject, method, expr.typeArgs, expr.args);
                    this.resolveReturnType(result, GenericsResolver.fromObject(resolvedObject));
                    return result;
                }
                else if (objectType is AnyType) {
                    expr.setActualType(AnyType.instance);
                    return expr;
                }
                else { }
                return resolvedObject;
            }
        }
        
        public override bool canTransform(Expression expr)
        {
            return expr is UnresolvedMethodCallExpression unrMethCallExpr && !(unrMethCallExpr.actualType is AnyType);
        }
        
        public override Expression transform(Expression expr)
        {
            return this.transformMethodCall(((UnresolvedMethodCallExpression)expr));
        }
    }
}